vllm/v1/metrics/perf.py

```
@@ -0,0 +1,939 @@
+        )
+
+    def get_num_flops_breakdown(
+        self, ctx: ExecutionContext, per_gpu: bool = True
+    ) -> Breakdown:
+        """Calculate flops breakdown for unembedding layer."""
+        D, V = self.hidden_size, self.vocab_size
+        T = ctx.num_tokens
+
+        if per_gpu:
+            V //= self.tp_size
+
+        flops = Breakdown()
+        flops.unembed = 2 * T * D * V
+
+        return flops
+
+    def get_read_bytes_breakdown(
+        self, ctx: ExecutionContext, per_gpu: bool = True
+    ) -> Breakdown:
+        """Calculate read memory traffic for unembedding layer."""
+        D, V = self.hidden_size, self.vocab_size
+        T = ctx.num_tokens
+
+        if per_gpu:
+            V //= self.tp_size
+
+        read_bytes = Breakdown()
+
+        read_bytes.input = T * D * self.activation_byte_size
+        read_bytes.weight = D * V * self.weight_byte_size
+
+        return read_bytes
+
+    def get_write_bytes_breakdown(
+        self, ctx: ExecutionContext, per_gpu: bool = True
+    ) -> Breakdown:
+        """Calculate write memory traffic for unembedding layer."""
+        V = self.vocab_size
+        T = ctx.num_tokens
+
+        if per_gpu:
+            V //= self.tp_size
+
+        write_bytes = Breakdown()
+
+        write_bytes.output = T * V * self.activation_byte_size
+
+        return write_bytes
+
+
+#### ModelMetrics ####
+
+
+class ModelMetrics:
+    def __init__(self, vllm_config: VllmConfig | None = None) -> None:
+        """
+        Parse vllm_config to instantiate metrics for each component.
+        Will raise ValueError if we cannot instantiate any components.
+        """
+
+        self.vllm_config = vllm_config
+
+        self.metrics: list[ComponentMetrics] = []
+        for metric_cls in ComponentMetrics.registered_metrics():
+            try:
+                metric = metric_cls.from_vllm_config(vllm_config)
+                self.metrics.append(metric)
+            except ValueError:
```

I'm wondering whether debug logging would be valuable here - e.g. if you were extending or adding a component, this would currently just swallow a useful error